src/cost_estimates/fee_scalar.rs

```diff
@@ -0,0 +1,207 @@
+        let db = Connection::open_with_flags(p, rusqlite::OpenFlags::SQLITE_OPEN_READ_WRITE)
+            .or_else(|e| {
+                if let SqliteError::SqliteFailure(ref internal, _) = e {
+                    if let rusqlite::ErrorCode::CannotOpen = internal.code {
+                        let mut db = Connection::open(p)?;
+                        let tx = tx_begin_immediate_sqlite(&mut db)?;
+                        Self::instantiate_db(&tx)?;
+                        tx.commit()?;
+                        Ok(db)
+                    } else {
+                        Err(e)
+                    }
+                } else {
+                    Err(e)
+                }
+            })?;
+        Ok(Self {
+            db,
+            metric,
+            decay_rate: 0.5_f64,
+        })
+    }
+
+    fn instantiate_db(c: &SqlTransaction) -> Result<(), SqliteError> {
+        c.execute(CREATE_TABLE, rusqlite::NO_PARAMS)?;
+        Ok(())
+    }
+
+    fn update_estimate(&mut self, new_measure: FeeRateEstimate) {
+        let next_estimate = match self.get_rate_estimates() {
+            Ok(old_estimate) => {
+                // compute the exponential windowing:
+                // estimate = (a/b * old_estimate) + ((1 - a/b) * new_estimate)
+                let prior_component = old_estimate.clone() * self.decay_rate;
+                let next_component = new_measure.clone() * (1_f64 - self.decay_rate);
+                let mut next_computed = prior_component + next_component;
+
+                // because of integer math, we can end up with some edge effects
+                // when the estimate is < decay_rate_fraction.1, so just saturate
+                // on the low end at a rate of "1"
+                next_computed.fast = cmp::max(next_computed.fast, 1);
+                next_computed.medium = cmp::max(next_computed.medium, 1);
+                next_computed.slow = cmp::max(next_computed.slow, 1);
+
+                next_computed
+            }
+            Err(EstimatorError::NoEstimateAvailable) => new_measure.clone(),
+            Err(e) => {
+                warn!("Error in fee estimator fetching current estimates"; "err" => ?e);
+                return;
+            }
+        };
+
+        debug!("Updating fee rate estimate for new block";
+               "new_measure_fast" => new_measure.fast,
+               "new_measure_medium" => new_measure.medium,
+               "new_measure_slow" => new_measure.slow,
+               "new_estimate_fast" => next_estimate.fast,
+               "new_estimate_medium" => next_estimate.medium,
+               "new_estimate_slow" => next_estimate.slow);
+
+        let sql = "INSERT OR REPLACE INTO scalar_fee_estimator
+                     (estimate_key, fast, medium, slow) VALUES (?, ?, ?, ?)";
+
+        let tx = tx_begin_immediate_sqlite(&mut self.db).expect("SQLite failure");
+
+        tx.execute(
+            sql,
+            rusqlite::params![
```

Comment:
Ooooh, new macro for `&[&dyn ToSql]`?

Reply:
Old macro, though we don't use it much, we should, because it's a lot nicer than peppering &dyn ToSqls